pennylane_qiskit/qiskit_device.py

```diff
@@ -180,61 +166,122 @@ def __init__(self, wires, provider, backend, shots=1024, **kwargs):
+            wires = operation.wires
+            par = operation.parameters
+            operation = operation.name
 
-        dag = circuit_to_dag(QuantumCircuit(self._reg, self._creg, name=""))
-        gate = mapped_operation(*par)
+            mapped_operation = self._operation_map[operation]
+
+            # TODO: Once a fix is available in Qiskit-Aer, remove the
+            # following:
+            if any(isinstance(x, np.ndarray) for x in par):
+                par = list(x if not isinstance(x, np.ndarray) else x.tolist() for x in par)
+
+            self.qubit_unitary_check(operation, par, wires)
+            self.qubit_state_vector_check(operation, par, wires)
+
+            qregs = [self._reg[i] for i in wires]
+
+            if operation in ("QubitUnitary", "QubitStateVector"):
+                # Need to revert the order of the quantum registers used in
+                # Qiskit such that it matches the PennyLane ordering
+                qregs = list(reversed(qregs))
+
+            dag = circuit_to_dag(QuantumCircuit(self._reg, self._creg, name=""))
+            gate = mapped_operation(*par)
+
+            if operation.endswith(".inv"):
+                gate = gate.inverse()
+
+            dag.apply_operation_back(gate, qargs=qregs)
+            circuit = dag_to_circuit(dag)
+            circuits.append(circuit)
+
+        return circuits
+
+    def qubit_state_vector_check(self, operation, par, wires):
+        """Input check for the the QubitStateVector operation."""
+        if operation == "QubitStateVector":
+            if self.backend_name == "unitary_simulator":
+                raise QuantumFunctionError(
+                    "The QubitStateVector operation "
+                    "is not supported on the unitary simulator backend."
+                )
 
-        if operation.endswith(".inv"):
-            gate = gate.inverse()
+            if len(par[0]) != 2 ** len(wires):
+                raise ValueError("State vector must be of length 2**wires.")
 
-        dag.apply_operation_back(gate, qargs=qregs)
-        qc = dag_to_circuit(dag)
-        self._circuit = self._circuit + qc
+    @staticmethod
+    def qubit_unitary_check(operation, par, wires):
+        """Input check for the the QubitUnitary operation."""
+        if operation == "QubitUnitary":
+            if len(par[0]) != 2 ** len(wires):
+                raise ValueError(
+                    "Unitary matrix must be of shape (2**wires,\
+                        2**wires)."
+                )
 
     def compile(self):
         """Compile the quantum circuit to target
         the provided compile_backend. If compile_backend is None,
         then the target is simply the backend."""
         compile_backend = self.compile_backend or self.backend
         compiled_circuits = transpile(self._circuit, backend=compile_backend)
+
+        # Specify to have a memory for hw/hw simulators
+        memory = str(compile_backend) in self._hw_backends
```

Comment:
This is so since the [`qasm_simulator` supports `memory`](https://github.com/Qiskit/qiskit-aer/blob/7de864cc989b354d0ad1febbd62b891270927b6b/qiskit/providers/aer/backends/qasm_simulator.py#L231).

Reply:
Hmm, wondering if it is better to have 
```suggestion
        memory = str(compile_backend) not in self._state_backends
```
and not explicitly define `hw_backends`. This avoids issues where the plugin needs to be updated as new hardware backends come online.

Reply:
As a matter of fact, I've added this option first. I think that's a fair point overall, it also assumes that the new backend would support `memory` which seems fair for a hardware backend. One downside is that it makes the code less explicit and we would by default support new backends without checking them (e.g. a situation might arise when that our test suite does not pick up a case related to a specific backend).

Having said that I'm on board with the change because it does create lower engineering overhead.